Add regression E2E test for the classic block initialization issue

fullofcaffeine opened the pull request:
## Description

Adds a regression test in the client block E2E spec to avoid the regression of this issue:  https://github.com/WordPress/gutenberg/issues/24696, as suggested in https://github.com/WordPress/gutenberg/pull/25162#issuecomment-689074703.

## How has this been tested?

- Check out this branch, and apply the following diff (effectively reverting #25162 and #25163):

```diff
diff --git a/packages/block-library/src/classic/edit.js b/packages/block-library/src/classic/edit.js
index c8da3b1b6d..8195e51006 100644
--- a/packages/block-library/src/classic/edit.js
+++ b/packages/block-library/src/classic/edit.js
@@ -54,11 +54,7 @@ export default class ClassicEdit extends Component {
                if ( document.readyState === 'complete' ) {
                        this.initialize();
                } else {
-                       document.addEventListener( 'readystatechange', () => {
-                               if ( document.readyState === 'complete' ) {
-                                       this.initialize();
-                               }
-                       } );
+                       window.addEventListener( 'DOMContentLoaded', this.initialize );
                }
        }
 
@@ -74,7 +70,7 @@ export default class ClassicEdit extends Component {
                } = this.props;
 
                const editor = window.tinymce.get( `editor-${ clientId }` );
-               const currentContent = editor?.getContent();
+               const currentContent = editor.getContent();
 
                if (
                        prevProps.attributes.content !== content &&

```

- Run `npm build` and `npm run wp-env-start`.
- Make sure the Gutenberg plugin is active in your local test instance. That's `localhost:8889` -- log into wp-admin to verify (credentials: `admin`/`password`)
- Run `npm run test-e2e -- packages/e2e-tests/specs/editor/blocks/classic.test.js`. (To watch tests in interactive mode, append `--puppeteer-interactive`).
- Verify that the `Classic › Should not fail after save/reload` test fails with `TypeError: Cannot read property 'getContent' of null`.
- Now remove the patch (`git checkout -- packages/block-library/src/classic/edit.js`), and rebuild (`npm run build`).
- Re-run the e2e test. It should pass this time.

## Types of changes

New E2E test.

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

fullofcaffeine committed:
Add regression E2E test for the classic block initialization issue

Issue: https://github.com/WordPress/gutenberg/issues/24696

sirreal commented:
This is failing for me now with:

```

 FAIL  packages/e2e-tests/specs/editor/blocks/classic.test.js (29.11s)
  Classic
    ✕ Should not fail after save/reload (16216ms)
    ○ skipped should be inserted
    ○ skipped should insert media, convert to blocks, and undo in one step

  ● Classic › Should not fail after save/reload

    TimeoutError: Element .edit-post-more-menu [aria-label="Options"] not found

      waiting for function failed: timeout 1300ms exceeded

```

ockham commented:
Some notes:

The e2e test can be run locally via `npm run wp-env start`, followed by `npm run test-e2e -- packages/e2e-tests/specs/editor/blocks/classic.test.js`. To watch them in interactive mode, append `--puppeteer-interactive` to the latter command.

I can repro the [issue](https://github.com/WordPress/gutenberg/pull/25169#issuecomment-689414551) Jon is seeing. Apparently, switching between Visual and Code Editor isn't working.

This test _should_ fail (earlier) if we revert 65c9f7477cc3a248f08841f64ef47aa490390b42 (that's #25162), _and_ run `npm run build`. However, it doesn't seem to :thinking: 

This is a screenshot I took while running `npm run test-e2e -- packages/e2e-tests/specs/editor/blocks/classic.test.js --puppeteer-interactive`:
 
![image](https://user-images.githubusercontent.com/96308/92633033-81071380-f2d2-11ea-8b08-f72ab11d8764.png)

It appears that the Gutenberg plugin isn't active during this e2e test, and that we're thus testing against the editor version that comes with Core, rather than the current GB codebase :thinking: I wonder if that's because we're running this test in isolation (and missing some 'global' setup function that enables the plugin), or whether this is a glitch with wp-env on Linux (which both @fullofcaffeine and I are using), or whether this currently affects all e2e tests :confused:

ockham commented:
Update, as Riad explained to me in Slack, the Gutenberg plugin has to be activated manually in the WP instance used for testing (at `localhost:8889`, credentials: admin/password).

ockham commented:
We've found that we need to disable Puppeteer cache before reloading the editor in order to repro the issue. The following works:

```diff
diff --git a/packages/e2e-tests/specs/editor/blocks/classic.test.js b/packages/e2e-tests/specs/editor/blocks/classic.test.js
index ce450cd172..d2e9ee06df 100644
--- a/packages/e2e-tests/specs/editor/blocks/classic.test.js
+++ b/packages/e2e-tests/specs/editor/blocks/classic.test.js
@@ -131,6 +131,7 @@ describe( 'Classic', () => {
                await saveDraft();
 
                // reload
+               await page.setCacheEnabled( false );
                await page.reload();
 
                await clickClassic();
```

We need to investigate if that turns off caching for all subsequent tests, and if yes, how we can restore the previous caching state.

fullofcaffeine committed:
Simplify classic block initialization issue regression test scenario

Considering commit 65c9f74 is reverted, the only actions needed to reproduce the issue are to:
 * Add the block to the editor, add some content;
 * Save;
 * Reload the editor;
 * Click it. It should error.

fullofcaffeine commented:
As @ockham, by disabling the cache, the regression is reproducible. It's not if the cache is left enabled (default).

You can test it by reverting the commit `65c9f74`, building GB again, and running this specific test.

I've chosen to go the simpler route of disabling, running the relevant steps to that reproduce the regression, then enabling the cache again. I assume it should work fine unless we have some kind of page object leakage in parallel tests, but I'm assuming that's not the case (I also don't know much about the Gutenberg testing infrastructure yet).

If anyone knows a better way to solve this or why browser caching causes the regression not to manifest in the E2E testing environment, please share your thoughts, we'd be keen to better understand why disabling the cache is needed in this case.

ockham commented on packages/e2e-tests/specs/editor/blocks/classic.test.js:
```diff
@@ -98,4 +101,45 @@ describe( 'Classic', () => {
+		// in case it regresses. To test this, revert commit 65c9f74, build and run the test.
+		await runWithoutCache( async () => {
+			await page.reload();
+			await clickClassic();
+			expect( console ).not.toHaveErrored();
+			expect( await getEditedPostContent() ).toMatchSnapshot();
+		} );
```
Do all of these need to be inside of `runWithoutCache`, or would just `await page.reload();` suffice?

```suggestion
		} );
		await clickClassic();
		expect( console ).not.toHaveErrored();
		expect( await getEditedPostContent() ).toMatchSnapshot();
```

fullofcaffeine replied:
Good point, I think only `page.reload()` is needed there. 👍🏻

fullofcaffeine committed:
Move non-relevant lines out of the `runWithoutCache`

Co-authored-by: Bernie Reiter <ockham@raz.or.at>

fullofcaffeine commented:
After https://github.com/WordPress/gutenberg/pull/25163 has been merged, if you want to reproduce this issue, you also need to revert it, in addition to `65c9f74`.

fullofcaffeine committed:
Remove obsolete snapshot

fullofcaffeine committed:
Remove unused import

ockham commented on packages/e2e-tests/specs/editor/blocks/classic.test.js:
```diff
@@ -46,6 +47,7 @@ describe( 'Classic', () => {
 		// Click the image button.
 		await page.waitForSelector( 'div[aria-label^="Add Media"]' );
 		await page.click( 'div[aria-label^="Add Media"]' );
+
```
```suggestion
```

ockham commented on packages/e2e-tests/specs/editor/blocks/classic.test.js:
```diff
@@ -98,4 +100,43 @@ describe( 'Classic', () => {
+			await page.click( mainBlockCSSSelector );
+		};
+
+		// Might move to utils if this becomes useful enough for other tests
+		const runWithoutCache = async ( cb ) => {
+			await page.setCacheEnabled( false );
+			await cb();
+			await page.setCacheEnabled( true );
+		};
```
TBH I was considering just inlining this, since it's really just two (fairly intuitive) commands. Might not make much sense to encapsulate them.

sirreal replied:
We need to make sure that `page.setCacheEnabled( true )` runs at the end even if the first bits throw.

ockham replied:
Ah, good point!

fullofcaffeine replied:
Ah yes, good point indeed!

fullofcaffeine replied:
> TBH I was considering just inlining this

Yeah, fine with me, we don't really use it anywhere else so you have a point.

fullofcaffeine replied:
Besides the actual method to disable and enable the cache are self-descriptive, the additional helper function doesn't add much in terms of meaning.

ockham committed:
Rename selector

ockham committed:
Use focus instead of click, inline

ockham commented:
I updated the PR desc with some step-by-step instructions.

ockham committed:
Add exception handling

fullofcaffeine commented:
> I updated the PR desc with some step-by-step instructions.

Thanks!

ockham commented:
Static Analysis job is failing. Fix: https://github.com/WordPress/gutenberg/pull/25223

ockham commented:
#25223 has been merged. I'll rebase this one.

ockham reviewed (approved):
Thanks a lot for working on this, @fullofcaffeine! Let's merge this once all checks are green :+1:
